File: csharp/triangle/Triangle.cs
    using System;

    public class Triangle
    {
        public static bool IsEquilateral(double a, double b, double c)
        {
            return !IsIllegal(a, b, c) && (a == b) && (b == c);
        }

        public static bool IsIsosceles(double a, double b, double c)
        {
            return !IsIllegal(a, b, c) && ((a == b) || (a == c) || (b == c));
        }

        public static bool IsScalene(double a, double b, double c)
        {
            return !IsIllegal(a, b, c) && (a != b) && (b != c);
        }

        private static bool IsIllegal(double a, double b, double c)
        {
            return (a + b < c) || (a + c < b) || (b + c < a) || (a + b + c == 0);
        }
    }

File: javascript/triangle/triangle.js
class Triangle {
    constructor(a, b, c) {
        this.a = a;
        this.b = b;
        this.c = c;
     }

    kind() {
        if (this._isIllegal()) {
            throw new Error();
        }
        
        if (this._isDegenerate()) {
            return "degenerate";
        }

        if (this._isEquilateral()) {
            return "equilateral";
        }
        
        if (this._isIsosceles()) {
            return "isosceles";
        }
        
        return "scalene";
    }

    _isIllegal() {
        return (this.a + this.b < this.c) 
            || (this.a + this.c < this.b) 
            || (this.b + this.c < this.a)
            || (this.a + this.b + this.c === 0);
    }

    _isDegenerate() {
        return (this.a + this.b === this.c) 
            || (this.a + this.c === this.b)
            || (this.b + this.c === this.a);
    }

    _isEquilateral() {
        return (this.a === this.b) && (this.b === this.c);
    }

    _isIsosceles() {
        return (this.a === this.b) || (this.b === this.c) || (this.a === this.c);
    }
}

module.exports = Triangle;
